consensus/safety-rules/src/safety_rules.rs

```diff
@@ -352,30 +350,34 @@ impl SafetyRules {
     }
 
     fn guarded_sign_proposal(&mut self, block_data: BlockData) -> Result<Block, Error> {
-        let mut safety_data = self.persistent_storage.safety_data()?;
-        self.signer()?;
+        let validator_signer = self.signer()?.clone();
         self.verify_author(block_data.author())?;
+
+        let mut safety_data = self.persistent_storage.safety_data()?;
         self.verify_epoch(block_data.epoch(), &safety_data)?;
+
         if block_data.round() <= safety_data.last_voted_round {
             return Err(Error::InvalidProposal(format!(
                 "Proposed round {} is not higher than last voted round {}",
                 block_data.round(),
                 safety_data.last_voted_round
             )));
         }
+
         self.verify_qc(block_data.quorum_cert())?;
         if self.verify_and_update_preferred_round(block_data.quorum_cert(), &mut safety_data)? {
             self.persistent_storage.set_safety_data(safety_data)?;
         }
 
         Ok(Block::new_proposal_from_block_data(
             block_data,
-            self.signer()?,
+            &validator_signer,
         ))
     }
 
     fn guarded_sign_timeout(&mut self, timeout: &Timeout) -> Result<Ed25519Signature, Error> {
-        self.signer()?;
+        let validator_signer = self.signer()?.clone();
```

Comment:
I'm confused at how this can work since signer shouldn't be cloneable. Can you try building without testing / fuzzing enabled.

Reply:
I think it clones the reference

Reply:
@davidiw, yeah, without looking that closely it clones the reference as self.signer()? returns an `&ValidatorSigner`. Thinking about it, I've yet to decide how bad this is.. 😺

Reply:
I don't understand what that means.

Reply:
A clone on a reference should return a copy of the value afaiu

Reply:
@davidiw, hmm, that's not what I see. For example, see here (https://www.reddit.com/r/rust/comments/d3wo9i/cloning_a_reference/) and the link to the rust code where '&' implements Clone() (https://doc.rust-lang.org/std/primitive.reference.html#trait-implementations

Reply:
Either way, this is still broken. Give me some time to think about the best way to do what I'd like to and I'll get back to you both 💩

Reply:
Maybe looking at the implementation will make it clearer?

https://doc.rust-lang.org/src/core/clone.rs.html#225-230

`&T::clone() -> T`

Reply:
Hah! Thanks @davidiw! 😃 Who would have thought that reddit wasn't the best place to get an authoritative opinion on rust...? 🤦 Anyway, I've just reverted the breaking nit picks. Overall the changes I'd have to make to get this to work are not worth it. Even as it is, this PR is a bit nit picky :P

Reply:
I have no comment on reddit or your nitpicking